File: 5-Problem-A-Day/250. Count Univalue Subtrees.java
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode left; TreeNode
 * right; TreeNode(int x) { val = x; } }
 */
class Solution {
    int count = 0;
    String diff = "";

    public int countUnivalSubtrees(TreeNode root) {
        count(root);
        return count;

    }

    private String count(TreeNode root) {
        if (root == null)
            return "null";
        String left = count(root.left), right = count(root.right);
        if ((left == "null" && right == "null")
                || (left == "null" && right.equals(Integer.toString(root.val)))
                || (right == "null" && left.equals(Integer.toString(root.val)))
                || (left.equals(right) && left.equals(Integer.toString(root.val)))) {
            count += 1;
            return Integer.toString(root.val);
        } else {
            diff += "x";
            return diff;
        }
    }
}

File: 2021/250. Count Univalue Subtrees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int count = 0;
    int countUnivalSubtrees(TreeNode* root) {
        if (root == NULL) return 0;
        getValSubtrees(root);
        return count;
    }
    
    int getValSubtrees(TreeNode* root) {                
        int left = root->left == NULL ? root->val : getValSubtrees(root->left);        
        int right = root->right == NULL ? root->val : getValSubtrees(root->right);        
        if (left == root->val && right == root->val) {
            count += 1;
            return root->val;
        }
        return 1001;
    }
};